neuralpp/symbolic/general_normalizer.py

```diff
@@ -0,0 +1,121 @@
+from .normalizer import Normalizer
+from .context_simplifier import ContextSimplifier
+from .expression import Expression, Constant, Variable, FunctionApplication, QuantifierExpression, Context, \
+    AbelianOperation
+from .z3_expression import Z3SolverExpression
+from .constants import basic_true, basic_false, if_then_else
+from .parameters import sympy_evaluate
+from .basic_expression import BasicQuantifierExpression, BasicExpression
+import neuralpp.symbolic.functions as functions
+from typing import List
+
+
+def _normalize(expression: Expression, context: Z3SolverExpression) -> Expression:
+    """
+    The function assumes context is satisfiable, otherwise will raise.
+    The input expression can be arbitrarily structured. The result expression is guaranteed to be
+    `quantifiers-at-leaves`: quantifiers are at leaves and do not contain conditional function in their bodies.
+    """
+    match expression:
+        case Constant():
+            return expression
+        case FunctionApplication(function=Constant(value=functions.conditional),
+                                 arguments=[condition, then, else_]):
+            if context.is_known_to_imply(condition):
+                return _normalize(then, context)
+            elif context.is_known_to_imply(~condition):
+                return _normalize(else_, context)
+            else:
+                return if_then_else(condition,
+                                    _normalize(then, context & condition),
+                                    _normalize(else_, context & ~condition))
+        case Expression(type=type_) if type_ == bool:
```

Comment:
Do you mean we also have `case Expression(type=type_) if type_ == bool:` there? I believe that's not quite correct because we only want to overload `__bool__()` for `==` symbolic expressions (and `Constant(True)` & `Constant(False)` in case `==` is simplified)

Reply:
Nah I was thinking if we can rewrite it into `case Constant(value=value) if isinstance(value, bool)`, but that's not super important. :)